walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,137 @@
+        List <TimeRange> finalTimes = new ArrayList<>();
+        finalTimes = sorter(finalTimes, request, newTimes);
+        // sort and filter times that optional attendees are available for
+        List <TimeRange> finalOptTimes = new ArrayList<>();
+        finalOptTimes = sorter(finalOptTimes, request, optTimes);
+
+        // check to see if the lists need to be compared; if there are no optional attendees, can
+        // return times that work for required attendees, and vice versa
+        if (request.getAttendees().size() == 0) {
+            return finalOptTimes;
+        } else if (request.getOptionalAttendees().size() == 0) {
+            return finalTimes;
+        }
+
+        // iterate through the optional attendees
+        List<TimeRange> possibleFinalTimes = new ArrayList<>();
+        possibleFinalTimes = comparer(finalOptTimes, finalTimes, request);
+
+        if (possibleFinalTimes.size() != 0) {
+            return possibleFinalTimes;
+        }
+        return finalTimes;
+    }
+
+    public Set<TimeRange> helper(Collection<Event> events, MeetingRequest request, Collection<String> attendees, Set<TimeRange> newTimes) {
+        for (String person: attendees) {
+            // iterate through the events already scheduled for that day
+            for (Event mtg: events) {
+                Set<String> curAttending = mtg.getAttendees();
+                TimeRange curWhen = mtg.getWhen();
+                if (curAttending.contains(person)) {
+                    List<TimeRange> times = new ArrayList<>();
+                    times.addAll(newTimes);
+                    // iterate through the available times
+                    for (TimeRange time: times) {
+                        if (time.contains(curWhen)) {
+                            newTimes.remove(time);
+                            newTimes.add(TimeRange.fromStartEnd(time.start(), curWhen.start(), false));
+                            newTimes.add(TimeRange.fromStartEnd(curWhen.end(), time.end(), false));
+                        } else if (time.overlaps(curWhen)) {
+                            if (time.contains(curWhen.start())) {
+                                newTimes.remove(time);
+                                newTimes.add(TimeRange.fromStartEnd(time.start(), curWhen.start(), false));
+                            } else if (curWhen.contains(time.start())) {
+                                newTimes.remove(time);
+                                newTimes.add(TimeRange.fromStartEnd(curWhen.end(), time.end(), false));
+                            }
+                        }
+                        if (curWhen.contains(time)) {
+                            newTimes.remove(time);
+                        }
+                    }
+                }
+            }
+        }
+        return newTimes;
+    }
+
+    public List<TimeRange> sorter(List<TimeRange> finalTimes, MeetingRequest request, Set<TimeRange> semiFilteredTimes) {
```

Comment:
Perhaps a comparator would make things easier? And then use a data structure that will keep it sorted as you add